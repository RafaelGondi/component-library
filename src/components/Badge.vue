<template>
	<div
		class="badge-container px-3 d-flex justify-content-center"
		:style="colorCodeMode ? styleVariables : ''"
		:class="variantMode ? predefinedStyle : ''"
	>
		<small class="regular">
			{{ content }}
		</small>
	</div>
</template>

<script>

export default {
	props: {
		content: {
			type: String,
			default: 'Negativo',
			description: 'The text that will be displayed inside the badge.',
			required: true,
		},
		color: {
			type: String,
			default: '#28A745',
			description: `If the color code mode is set,
				you can pass the hex value of the text
				color of the badge.`,
		},
		bgColor: {
			type: String,
			default: '#28a74526',
			description: `If the color code mode is set,
				you can pass the hex value of the background
				color of the badge.`,
		},
		variantMode: {
			type: Boolean,
			default: true,
			description: `The property that specifies 
				if the badge will work in variant mode
				or not.`,
		},
		colorCodeMode: {
			type: Boolean,
			default: false,
			description: `The property that specifies 
				if the badge will work in the color code
				mode or not.`,
		},
		variant: {
			type: String,
			default: 'yellow',
			description: `Variant mode gives 6 predefined badges 
				to work with. Seting this property to 'true',
				you can use 'yellow', 'green', ... 'gray'
				to use the predefined badge styles.`,
		},
	},

	computed: {
		predefinedStyle() {
			let computed_style = '';
			switch (this.variant) {
				case 'yellow':
					computed_style = 'yellow';
					break;
				case 'green':
					computed_style = 'green';
					break;
				case 'red':
					computed_style = 'red';
					break;
				case 'blue':
					computed_style = 'blue';
					break;
				case 'purple':
					computed_style = 'purple';
					break;
				case 'gray':
					computed_style = 'gray';
					break;
				default:
					computed_style = 'yellow';
					break;
			}

			return computed_style;
		},

		styleVariables() {
			return {
				'--bg-color': this.bgColor,
				'--color': this.color,
			};
		},
	},
};
</script>
<style>
.badge-container {
	border-radius: 8px !important;
	width: fit-content;
	background-color: var(--bg-color);
	color: var(--color);
}

.yellow {
	color: #5a4300;
	background-color: #ffbf004d;
}

.green {
	color: #005a15;
	background-color: #28a74526;
}

.red {
	color: #8c1520;
	background-color: #e2757f40;
}

.blue {
	color: #00469c;
	background-color: #60c4ff40;
}

.purple {
	color: #621e83;
	background-color: #e47aff40;
}

.gray {
	color: #2f2f2f;
	background-color: #acacac40;
}
</style>
