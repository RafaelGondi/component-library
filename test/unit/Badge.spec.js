import { mount, createLocalVue } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';
import Badge from '../../src/components/Badge.vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

test('Component is mounted properly', () => {
	const wrapper = mount(Badge, {
		localVue,
		propsData: {
			content: 'Test',
		},
	});
	expect(wrapper).toMatchSnapshot();
});

describe("Computed property 'predefinedStyle' test", () => {
	test('if the computed property predefinedStyle works properly with the default value for the prop variant', () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				content: 'Test',
			},
		});
		expect(wrapper.vm.predefinedStyle).toBe('variant-1');
	});
});

describe("Prop 'variant' tests", () => {
	test("if the computed property changes when the prop variant is setted to 'variant-1'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'variant-1',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('variant-1');
	});

	test("if the computed property changes when the prop variant is setted to 'variant-2'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'variant-2',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('variant-2');
	});

	test("if the computed property changes when the prop variant is setted to 'variant-3'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'variant-3',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('variant-3');
	});

	test("if the computed property changes when the prop variant is setted to 'variant-4'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'variant-4',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('variant-4');
	});

	test("if the computed property changes when the prop variant is setted to 'variant-5'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'variant-5',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('variant-5');
	});

	test("if the computed property changes when the prop variant is setted to 'variant-6'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				variant: 'variant-6',
				content: 'Test',
			},
		});

		expect(wrapper.vm.predefinedStyle).toBe('variant-6');
	});
});

describe("Prop 'bgColor' and 'color' tests", () => {
	test("if the computed property 'styleVariables' changes when the prop 'colorCodeMode' is setted to 'true'", () => {
		const wrapper = mount(Badge, {
			localVue,
			propsData: {
				colorCodeMode: true,
				variantMode: false,
				color: 'white',
				bgColor: 'blue',
				content: 'Test',
			},
		});

		expect(wrapper.vm.styleVariables).toStrictEqual({"--bg-color": "blue", "--color": "white"});
	});
});
