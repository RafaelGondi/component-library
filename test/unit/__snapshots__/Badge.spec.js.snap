// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<div class="badge-container px-3 d-flex justify-content-center variant-1"><small class="regular">
    Test
  </small></div>
`;
